### PR TITLE
Service node list implementation

### PR DESCRIPTION
**Pull request by jcktm**

### CODE REVIEW

File: src/cryptonote_core/service_node_list.cpp [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+        if (!m_blockchain.get_transactions(block.tx_hashes, txs, missed_txs))
+        {
+          LOG_ERROR("Unable to get transactions for block " << block.hash);
+          return;
+        }
+        block_added_generic(block, txs);
+      }
+    }
+
+    m_rollback_events.push_back(new prevent_rollback(current_height));
```
Comment (Doy-lee) [inline src/cryptonote_core/service_node_list.cpp]: A unique_ptr would work here for C++11. `std::unique_ptr<rollback_event>(new prevent_rollback(current_height)) `which has the benefit of removing the delete calls.

File: src/cryptonote_core/blockchain.h [hunk 2]
```diff
@@ -286,6 +315,13 @@ namespace cryptonote
      */
     difficulty_type get_difficulty_for_next_block();
 
+    /**
+     * @brief returns the staking requirement for the block at height
+     *
+     * @return the target
+     */
+    uint64_t get_staking_requirement(uint64_t height);
```
Comment (Doy-lee) [inline src/cryptonote_core/blockchain.h]: This can be a const function.

File: src/cryptonote_core/service_node_list.cpp (hunk 1 above)
Comment (jcktm) [reply src/cryptonote_core/service_node_list.cpp]: Nice, thanks

File: src/cryptonote_core/blockchain.h (hunk 2 above)
Comment (jcktm) [reply src/cryptonote_core/blockchain.h]: Fixed

File: src/cryptonote_core/service_node_list.h [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    cryptonote::account_public_address select_winner(const crypto::hash& prev_id);
+
+    std::vector<cryptonote::account_public_address> get_service_nodes_pubkeys();
+
+  private:
+    bool process_registration_tx(const cryptonote::transaction& tx, uint64_t block_height, cryptonote::account_public_address& address);
+    template<typename T>
+    void block_added_generic(const cryptonote::block& block, const T& txs);
+
+    bool reg_tx_has_correct_unlock_time(const cryptonote::transaction& tx, uint64_t block_height);
```
[the remark this answers is not in the document]
Comment (jcktm) [reply src/cryptonote_core/service_node_list.h]: Fixed

**Pull request merged**